eZ/Bundle/EzPublishCoreBundle/Command/NormalizeImagesPathsCommand.php

```diff
@@ -209,9 +173,117 @@ private function moveFile(
             ]
         );
 
-        $newBinaryFile = $this->ioService->createBinaryFile($binaryCreateStruct);
-        if ($newBinaryFile instanceof BinaryFile) {
-            $this->ioService->deleteBinaryFile($oldBinaryFile);
+        // Before creating a new file validate if the same file doesn't exist already in order to not duplicate files
+        $newBinaryFile = $this->ioService->loadBinaryFileByUri(\DIRECTORY_SEPARATOR . $newPath);
+        if ($newBinaryFile instanceof MissingBinaryFile) {
+            $this->ioService->createBinaryFile($binaryCreateStruct);
+        }
+    }
+
+    protected function updateImagePathsToNormalize(
+        $imageData,
+        array $imagePathsToNormalize
+    ): array {
+        $filePath = $imageData['filepath'];
+        $fieldId = (int)$imageData['contentobject_attribute_id'];
+
+        $finalNormalizedPath = $this->getFinalNormalizedPath(
+            $filePath,
+            $imagePathsToNormalize
+        );
+
+        if ($finalNormalizedPath !== $filePath) {
+            $imagePathsToNormalize[$fieldId] = [
+                'fieldId' => $fieldId,
+                'oldPath' => $filePath,
+                'newPath' => $finalNormalizedPath,
+            ];
         }
+
+        return $imagePathsToNormalize;
+    }
+
+    private function getFinalNormalizedPath(
+        string $filePath,
+        array $imagePathsToNormalize
+    ): string {
+        $processedPaths = array_values(
+            array_filter(
+                $imagePathsToNormalize,
+                static function (array $data) use ($filePath) {
+                    return $data['oldPath'] === $filePath;
+                }
+            )
+        );
+
+        return !empty($processedPaths)
+            ? $processedPaths[0]['newPath']
+            : $this->filePathNormalizer->normalizePath($filePath);
+    }
+
+    private function getImagePathsToNormalize(SymfonyStyle $io): array
+    {
+        $imagesCount = $this->imageGateway->countDistinctImages();
+        $imagePathsToNormalize = [];
+        $iterations = ceil($imagesCount / self::IMAGE_LIMIT);
+        $io->progressStart($imagesCount);
+        for ($i = 0; $i < $iterations; ++$i) {
+            $imagesData = $this->imageGateway->getImagesData(
+                $i * self::IMAGE_LIMIT,
+                self::IMAGE_LIMIT
+            );
+
+            foreach ($imagesData as $imageData) {
+                $imagePathsToNormalize = $this->updateImagePathsToNormalize(
+                    $imageData,
+                    $imagePathsToNormalize
+                );
+
+                $io->progressAdvance();
+            }
+        }
```

Comment:
Ensure that this works properly on Postgres. My own testing for an unrelated feature with a similar batch processing mechanism has shown that due to how Postgresql handles updates, this non-sorted query might result in records being omitted or yielded more than once.
This is because for Postgres an update operation moves the record object last (or the order becomes otherwise undetermined). Sorting the query in gateways that handle it by primary key should be the easiest solution.

Reply:
Noted in the ticket for QA.